File: packages/oxlint-plugin-react-doctor/src/plugin/utils/get-static-object-property-value.ts
import type { EsTreeNode } from "./es-tree-node.js";
import { getStaticPropertyKeyName } from "./get-static-property-key-name.js";
import { isNodeOfType } from "./is-node-of-type.js";
import { stripParenExpression } from "./strip-paren-expression.js";

export const getStaticObjectPropertyValue = (
  node: EsTreeNode,
  expectedPropertyName: string,
): EsTreeNode | null | undefined => {
  const expression = stripParenExpression(node);
  if (!isNodeOfType(expression, "ObjectExpression")) return null;
  let propertyValue: EsTreeNode | null | undefined;
  for (const property of expression.properties) {
    if (!isNodeOfType(property, "Property")) {
      propertyValue = null;
      continue;
    }
    const propertyName = getStaticPropertyKeyName(property, { allowComputedString: true });
    if (propertyName === null) {
      propertyValue = null;
      continue;
    }
    if (propertyName !== expectedPropertyName) continue;
    propertyValue = property.kind === "init" ? property.value : null;
  }
  return propertyValue;
};

File: native/oxlint/rules/get-static-object-property-value.rs
fn get_static_object_property_value<'a>(
    expression: &'a oxc_ast::ast::Expression<'a>,
    expected_property_name: &str,
) -> Option<&'a oxc_ast::ast::Expression<'a>> {
    let oxc_ast::ast::Expression::ObjectExpression(object_expression) =
        expression.get_inner_expression()
    else {
        return None;
    };
    let mut property_value = None;
    for property in &object_expression.properties {
        let oxc_ast::ast::ObjectPropertyKind::ObjectProperty(property) = property else {
            property_value = None;
            continue;
        };
        let Some(property_name) = property.key.static_name() else {
            property_value = None;
            continue;
        };
        if property_name != expected_property_name {
            continue;
        }
        property_value =
            (property.kind == oxc_ast::ast::PropertyKind::Init).then_some(&property.value);
    }
    property_value
}
